Adds general article about Heartcore Grid Editors

rasmusjp opened the pull request:
I'm not entirely sure if it should be created under `API Documentation` but I alse feel like it doesn't fit under `Getting Started Cloud`.

Maybe we should have the `Backoffice` page at the root of `Umbrace Heartcore` and then move the backoffice specific articles from `Getting Started Cloud` to this instead?

The `Umbraco-Heartcore/API-Documentation/Backoffice/index.md` should probably be updated with some more text 😅 

The `Limitations and best practices` part is taken from https://github.com/umbraco/UmbracoDocs/blob/7ec94d58e0754da98a9c8f9af713bc5edcf1222a/Umbraco-Heartcore/Tutorials/Creating-A-Custom-Grid-Editor/index.md so we might want to remove it from that page and link to this article instead

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
````diff
@@ -0,0 +1,147 @@
+}
+```
+
+When importing libraries that registers variables or functions on the global object or custom elements it's advised to create a Module Alias and import that instead. This is to avoid conflicts with importing different versions of a library.
````
```suggestion
When importing libraries that register variables or functions on the global object or custom elements, it's advised to create a Module Alias and import that instead. This avoids conflict while importing different versions of a library.
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,147 @@
+
+![Screenshot showing the Settings section with the Grid Editors tree expanded](images/settings-grid-editors.png)
+
+The Grid Editor must export a default class inheriting from `HTMLElement`
```
```suggestion
The Grid Editor must export a default class inheriting from the `HTMLElement`
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,147 @@
+![Screenshot showing the Module Aliases configuration](images/module-aliases.png)
+
+:::tip
+Use a Module Alias when importing libraries to avoid conflicts by importing different versions
```
```suggestion
Use a Module Alias when importing libraries to avoid conflicts by importing different versions.
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,147 @@
+
+## JSON Schema
+
+A [JSON Schema](https://json-schema.org/) is used to describe the how the value is stored and returned by the REST and GraphQL API.
```
```suggestion
A [JSON Schema](https://json-schema.org/) is used to describe how the value is stored and returned by the REST and GraphQL API.
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,147 @@
+
+This allows for advanced configurations where for example an [UDI](https://our.umbraco.com/Documentation/Reference/Querying/UDI-identifiers/) is converted to a URL.
+
+The default Schema looks like this
```
```suggestion
The default Schema looks like this:
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
````diff
@@ -0,0 +1,147 @@
+
+```
+
+Since it has `"type": "string"` the API's will return the stored value as a string.
````
```suggestion
Since it has `"type": "string"` the APIs will return the stored value as a string.
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,147 @@
+
+Since it has `"type": "string"` the API's will return the stored value as a string.
+
+If the editor stores an `object` we can use `"type": "object"` instead, now the API's will return the value as a JSON object.
```
```suggestion
If the editor stores an `object` we can use `"type": "object"` instead, now the APIs will return the value as a JSON object.
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,147 @@
+
+If the editor stores an `object` we can use `"type": "object"` instead, now the API's will return the value as a JSON object.
+
+By also adding properties, the API will be more clever of how the data is returned.
```
```suggestion
By also adding properties, the API will be more clever in how the data is returned.
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,147 @@
+
+By also adding properties, the API will be more clever of how the data is returned.
+
+For example an image editor that stores an `UDI` in the `url` field and the alt text in the `altText` field like this
```
```suggestion
For example, an image editor that stores an `UDI` in the `url` field and the alt text in the `altText` field like this:
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
````diff
@@ -0,0 +1,147 @@
+}
+```
+
+Could have a schema like
````
```suggestion
Could have a schema like:
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
````diff
@@ -0,0 +1,147 @@
+}
+```
+
+Notice how there's a `"format":·"uri-reference"` to the `url` property in the schema. Combined with `"type": "string"` the API's knows this is a URI and will try to parse the value stored as an `UDI` and if successful, return the URL of the referenced item.
````
```suggestion
Notice how there's a `"format":·"uri-reference"` to the `url` property in the schema. Combined with `"type": "string"` the API knows this is a URI and will try to parse the value stored as an `UDI` and if successful, return the URL of the referenced item.
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,147 @@
+
+Notice how there's a `"format":·"uri-reference"` to the `url` property in the schema. Combined with `"type": "string"` the API's knows this is a URI and will try to parse the value stored as an `UDI` and if successful, return the URL of the referenced item.
+
+Currently the following combinations are supported:
```
```suggestion
Currently, the following combinations are supported:
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,147 @@
+
+## Limitations and best practices
+
+Currently there are some functionality and components that does not work well in the preview pane. This includes integrations with the backoffice like pickers and the Rich Text Editor. They do however still work when inserted on a page.
```
```suggestion
Currently, some functionality and components do not work well in the preview pane. This includes integrations with the backoffice like Pickers and the Rich Text Editor. However, they still work when inserted on a page.
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,147 @@
+
+To make your custom editors less likely to break with future updates, do not use any of the backoffice javascript directly. Always use the [headless backoffice bridge](https://github.com/umbraco/Umbraco.Headless.Backoffice.Bridge).
+
+If the library is missing any functionality raise an issue on the [Heartcore issue tracker](https://github.com/umbraco/Umbraco.Heartcore.Issues/issues).
```
```suggestion
If the library is missing any functionality, raise an issue on the [Heartcore issue tracker](https://github.com/umbraco/Umbraco.Heartcore.Issues/issues).
```

eshanrnh commented on Umbraco-Heartcore/API-Documentation/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,147 @@
+
+If the library is missing any functionality raise an issue on the [Heartcore issue tracker](https://github.com/umbraco/Umbraco.Heartcore.Issues/issues).
+
+Try to avoid relying on backoffice CSS-classes. Instead we recommend creating isolated elements using [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM).
```
```suggestion
Try to avoid relying on backoffice CSS-classes. Instead, we recommend creating isolated elements using [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM).
```

rasmusjp commented:
Hi @eshanrnh 

Thanks, I'll have a look at the suggestions

> Regarding the placement of the article, you are right, it doesn't fit under API Documentation. Rather creating it under Backoffice will make more sense (something similar to https://our.umbraco.com/documentation/Fundamentals/Backoffice/).
> 
> I suppose once the Backoffice section has more articles/sub-articles, the Backoffice page can be updated with more text 😀

Yeah, I was thinking having a new section https://our.umbraco.com/documentation/Umbraco-Heartcore/Backoffice/, I'll move it. 

There are some articles under https://our.umbraco.com/documentation/Umbraco-Heartcore/Getting-Started-Cloud/ that are backoffice specific which I think also would make more sense to have under a Backoffice section, I can make that in another PR when this is merged.

> The Limitations and best practices part in the Creating-A-Custom-Grid-Editor Tutorial should reference a link to this section here 💪 Maintaining duplicate information will be challenging later on 🙂

I agree, but wasn't sure how to handle it, I just noticed the Styled links which fits perfectly here. I'll update the other tutorial with a link to this article

eshanrnh commented:
Great, if you need help/collaboration with your PR, feel free to reach out 😀

rasmusjp commented:
I've moved the pages and applied the suggestions (think I got them all), some of it I've rephrased.

I've also added links betweeen the article and the tutorial

eshanrnh commented on Umbraco-Heartcore/Backoffice/Grid-Editors/index.md:
````diff
@@ -0,0 +1,151 @@
+}
+```
+
+When importing libraries that registers variables or functions on the global object or custom elements it's advised to create a Module Alias and import that instead. By using a Module Alias when importing we can ensure that only one version of a library is imported.
````
```suggestion
When importing libraries that registers variables or functions on the global object or custom elements it's advised to create a Module Alias and import that instead. By using Module Alias when importing we can ensure that only one version of a library is imported.
```

eshanrnh commented on Umbraco-Heartcore/Backoffice/Grid-Editors/index.md:
```diff
@@ -0,0 +1,151 @@
+![Screenshot showing the Module Aliases configuration](images/module-aliases.png)
+
+:::tip
+Use a Module Alias when importing libraries to ensure only a single verson of that library is loaded.
```
```suggestion
Use a Module Alias when importing libraries to ensure only a single version of that library is loaded.
```

eshanrnh commented on Umbraco-Heartcore/Backoffice/index.md:
```diff
@@ -0,0 +1,13 @@
+
+# Working in the backoffice
+
+On this page you'll find articles about the backoffice that's specific to Heartcore.
```
```suggestion
On this page, you'll find articles about the backoffice that are specific to Heartcore.
```

eshanrnh reviewed (commented):
The changes look perfect, @rasmusjp 🙌 Just some final tweaks🙂

eshanrnh commented:
Thanks for accepting the suggestions, @rasmusjp 💪Merging the PR.
